Add `fetch_with_deferred_update` method to `ActiveSupport::Cache` for out-of-band cache updates

erikbrannstrom opened the pull request:
### Summary

The new option specifies that `Cache.fetch` will call the given block and then return the expired value from the cache. This allows cache updates to be performed out-of-band, which can be useful for computation heavy operations.

An alternative would be to continuously update the cached value using a scheduled task, but that decouples the logic and will make it hard to know where the update takes place.

The following code would for example update the cache every 6 hours. Once the cache value expires, the block will be called which starts a job that will eventually update the cache with a new value. During the grace period specified by `race_condition_ttl` the expired value will be returned for all calls to `fetch`.
```ruby
Rails.cache.fetch(cache_key, expires_in: 6.hours, defer_update: true, race_condition_ttl: 300) do
  HeavyComputationJob.perform_later cache_key
end
```

I believe this could be useful for more people than just me, and I'm looking forward to get the input of the community! Also, this is my first PR to Rails! I've followed along the contribution guide, but if I missed something I will try to fix it ASAP!

### Discussion / Questions
**Option naming**
I have struggled somewhat with the naming of the option. I think `defer_update` is fairly good, but it could potentially imply that the block itself will be magically called asynchronously. I'm open for suggestions on this one!

**`fetch_multi`**
I have not implemented this behavior for `fetch_multi` because I didn't feel it made as much sense, but one could argue it makes the methods somewhat asymmetric. Not sure if that is a problem though.

erikbrannstrom committed:
Add `:defer_update` option to `Cache.fetch`

The new option specifies that `Cache.fetch` will call the given
block and then return the expired value from the cache. This
allows cache updates to be performed out-of-band, which can be
useful for computation heavy operations.

rafaelfranca commented:
It is not clear to me the difference between

```
Rails.cache.fetch(cache_key, expires_in: 6.hours, defer_update: true, race_condition_ttl: 300) do
  HeavyComputationJob.perform_later cache_key
end
```

and

```
Rails.cache.fetch(cache_key, expires_in: 6.hours, race_condition_ttl: 300) do
  HeavyComputationJob.perform_later cache_key
end
```

Could you explain?

erikbrannstrom commented:
@rafaelfranca A valid question, and I could have made that more clear!

The difference is that in the first case (with the `defer_update` option), the `fetch` call will return the value that just expired from the cache, where as in the latter case, the return value will be whatever the block returns (in this case, an ActiveJob instance).

A more helpful example would perhaps be assigning the resulting value to an instance variable to be used in a view:

```ruby
@important_value = Rails.cache.fetch(cache_key, expires_in: 6.hours, defer_update: true, race_condition_ttl: 300) do
  HeavyComputationJob.perform_later cache_key
end
```

With the new option, `@important_value` will always get a cached value (unless the race condition TTL expires as well before a new value has been written to the cache), where as without it there will be one render where `@important_value` will not be available.

Did that clear things up?

casperisfine commented:
I wonder if it wouldn't make more sense as a distinct method rather than an option.

rafaelfranca commented:
Yes. I was about to comment that. I think I prefer a new method that implement the defer logic.

erikbrannstrom commented:
@casperisfine @rafaelfranca Yeah, I agree that the extra option might just add too much complexity to the `fetch` method.

Taking a step back, I would consider the following to be the requirements of a new method.
1) It should return the cache value just like `read` or `fetch` as long as it has not expired
2) If the cached value has expired, some user defined code should be called, which in turn is responsible for refreshing the cache
3) Since it may take some time to refresh the cache, it should be possible to use the expired cache value for a grace period and avoid refreshing the cache for every call made during this period

So how about something like this?

```ruby
@value = Rails.cache.fetch_with_refresh("key", max_refresh_time: 5.minutes) do |cache_key|
  HeavyComputationJob.perform_later cache_key
end
```

Once an entry has expired, the block will be called, the entry will have its expiry time set to `max_refresh_time` from now, and `@value` will be assigned the old cache value.

I'm not really loving the name `fetch_with_refresh` though, but I think _refresh_ may be a better word to use than _deferred update_. I used the `fetch` part of the name because the basic idea is still that the method should both read and update the value, even if the _how_ differs from the normal `fetch` method.

I'd love to hear your thoughts on this!

casperisfine commented:
> it should be possible to use the expired cache value for a grace period and avoid refreshing the cache for every call made during this period

IMHO it should actually be mandatory to set the `race_condition_ttl` on this new API. `max_refresh_time` might be more telling, but I think `race_condition_ttl` would make it clearer than it's the same behavior than for `fetch`.

> I'm not really loving the name fetch_with_refresh

Me neither 😅 . I don't have any good idea coming to mind right now :/

erikbrannstrom commented:
> IMHO it should actually be mandatory to set the `race_condition_ttl` on this new API. `max_refresh_time` might be more telling, but I think `race_condition_ttl` would make it clearer than it's the same behavior than for `fetch`.

I'm not necessarily against using `race_condition_ttl`, but my initial thought was a slightly different behavior than that option. For this case, I'm thinking each time the block is called, the cached value simply gets a new expiry time. If I read the code right (which is a semi-big _if_ TBH), that is how `race_condition_ttl` will work for most cache adapters, except for the Redis cache, where there seems to be a "hard expire" once the TTL is reached.

But as I said, it's not a super strong opinion, just my initial reasoning. I'm gonna sleep on it, and perhaps more importantly, sleep on the method name 🙈

rafaelfranca commented:
`fetch_and_defer_update`
`get_and_update_with`

are some ideas.

kaspth commented:
I like fetch_and_defer_update, could also do fetch_with_deferred_update 👍

natematykiewicz commented:
This sounds a lot like a use case I described at work today.

We have a monthly import that loads data from a 3rd party vendor into a few database tables.

We need some grouped counts of this data for a web page. Since the data literally only changes once a month, we were thinking we could use ActiveSupport cache to cache our changes, but that it'd be nice to be able to also use ActiveJob to refresh the cache periodically so that no user ever has to build the cache during a web request.

If I'm reading this PR right (I'm not familiar with race_condition_ttl), I love the idea.

erikbrannstrom committed:
Implement Cache#fetch_with_deferred_update

Added a new method, fetch_with_deferred_update, which allows for
out-of-band updates of cache values.

erikbrannstrom commented:
I've re-implemented the solution as a separate method. I ended up going with the name `fetch_with_deferred_update`.

I moved some shared logic in `fetch` and `fetch_with_deferred_update` to a private method called `fetch_entry`, which seemed somewhat analogous to e.g. `read_entry`. It is passed a symbol for the calling method's name, which is passed to instrumentation, and a named parameter to control whether expired values are be cleared or not.

Looking forward to getting your input on this! :+1:

casperisfine commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -350,6 +341,80 @@ def fetch(name, options = nil, &block)
+        if !entry
+          yield(name)
+        elsif entry.expired?
+          write(name, entry.value, options.merge(expires_in: additional_cache_time))
```
if you were to pass `race_condition_ttl: additional_cache_time ` to `fetch_entry`, you'd get the extension logic already.

erikbrannstrom replied:
@casperisfine Won't it be a slightly different behavior? I wrote the following comment a few days ago:

> I'm not necessarily against using race_condition_ttl, but my initial thought was a slightly different behavior than that option. For this case, I'm thinking each time the block is called, the cached value simply gets a new expiry time. If I read the code right (which is a semi-big if TBH), that is how race_condition_ttl will work for most cache adapters, except for the Redis cache, where there seems to be a "hard expire" once the TTL is reached.

Is my interpretation correct? If so, I'm wondering if it's better for this method to always do the same thing, no matter which adapter is used, rather than to opaquely use an option which is handled slightly differently between implementations.

What do you think?

casperisfine replied:
> except for the Redis cache, where there seems to be a "hard expire" once the TTL is reached.

Hum, I can't see such difference. What makes you think that?

erikbrannstrom replied:
@casperisfine My bad! 🙈 I had another look at the code and honestly can't remember what made me so sure...

erikbrannstrom replied:
@casperisfine Done :+1:

ghiculescu commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -350,6 +341,80 @@ def fetch(name, options = nil, &block)
+      # specified, and then the block will be executed in the same way as for a
+      # missing value.
+      #
+      # In contrary to `fetch`, the return value of the block will not be used by the
```
```suggestion
      # Unlike `fetch`, the return value of the block will not be used by the
```

ghiculescu commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -350,6 +341,80 @@ def fetch(name, options = nil, &block)
       end
 
+      # Fetches data from the cache, using the given key. If there is data in
+      # the cache with the given key, then that data is returned, otherwise +nil+
```
```suggestion
      # the cache with the given key and that data hasn't expired, then that data is returned, otherwise +nil+
```

is that right? or will it be returned even if it's expired?

erikbrannstrom replied:
Currently it would always be returned, but by changing the implementation to use `race_condition_ttl` it would have a slightly different behavior. I'll fix the comment with that change :+1:

ghiculescu commented on activesupport/CHANGELOG.md:
````diff
@@ -1,3 +1,13 @@
+*   Add `Cache#fetch_with_deferred_update` to allow out-of-band updates of cache values.
+
+    ```ruby
+    Rails.cache.fetch_with_deferred_update(cache_key, 5.minutes) do
````
what do you think about making this a keyword argument, so it's more explicit the purpose of it?

```suggestion
    Rails.cache.fetch_with_deferred_update(cache_key, refresh_grace_period: 5.minutes) do
```

I had to look at it a bit to understand, but I think the correct way to use the API is to have that value be roughly how long the block will take to complete (so it's the grace period during which the cache is being rebuilt). Is that right?

erikbrannstrom replied:
I went with `:race_condition_ttl` as it is already an existing option for `fetch` :+1:

erikbrannstrom committed:
Use :race_condition_ttl

Replace custom expiry logic with `race_condition_ttl` option.

erikbrannstrom commented:
Some additional tweaks! I went with @casperisfine's suggestion to use `:race_condition_ttl` instead of the custom expiry logic. Though it's not quite as nice for my specific use case, this makes new method much more aligned with how `fetch` works.

There is a difference though in that `fetch_with_deferred_update` will actually return the stale data in the call that also bumps the expiry time. This is because `fetch` is supposed to simply return the new value in the block, where as this method expects that calculation to be performed out-of-band and thus it makes more sense to let **all** calls during the `:race_condition_ttl` interval to get the stale value.

byroot reviewed (approved):
I'm not certain the name is that clear. The `_with_deferred_update` might suggest that somehow it's work exactly like fetch except that the block is somehow deferred on a background thread or something like that.

The API doesn't even force you to do an update in the block, you might just want to know that the entry is about to be expired (e.g. to show a warning to the user or something).

 I think it could have been `read_stale` or `fetch_stale`. Or even just `refresh`.

The implementation look good to me, hence why I'm approving, but I'd love another round of opinion on the naming from @kaspth and @rafaelfranca

rafaelfranca commented:
About the name I think `read_stale_and_defer_update` would be clear to me, but that seems too long.

It is true that nothing says you need to call a job inside the block. Only `read_stale` makes sense to me in that case.

erikbrannstrom commented on activesupport/lib/active_support/cache.rb:
```diff
@@ -599,6 +642,21 @@ def deserialize_entry(payload)
+            entry = handle_expired_entry(cached_entry, key, options, return_stale_entry: return_stale_entry)
+            entry = nil if entry && entry.mismatched?(normalize_version(name, options))
+            payload[:super_operation] = operation if payload
+            payload[:hit] = !!entry if payload
```
Good catch, I'll fix it once we've settled on a name for the method 😄

erikbrannstrom commented:
Good input on the naming! I'd be fine changing the name to `read_stale`, though that name could be interpreted as _only_ reading stale values. Perhaps `read_with_stale` would make that more clear?
